src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -234,222 +241,226 @@ void HRVOAgent::computeNewVelocity()
             }
 
-            d = (velocityObstacles_[i].side2_).determinant(velocityObstacles_[j].side2_);
+            d = (velocityObstacles_[i].left_side)
+                    .determinant(velocityObstacles_[j].left_side);
 
             if (d != 0.0f)
             {
                 const float s =
                     (velocityObstacles_[j].apex_ - velocityObstacles_[i].apex_)
-                        .determinant(velocityObstacles_[j].side2_) /
+                        .determinant(velocityObstacles_[j].left_side) /
                     d;
                 const float t =
                     (velocityObstacles_[j].apex_ - velocityObstacles_[i].apex_)
-                        .determinant(velocityObstacles_[i].side2_) /
+                        .determinant(velocityObstacles_[i].left_side) /
                     d;
 
                 if (s >= 0.0f && t >= 0.0f)
                 {
                     candidate.position_ =
-                        velocityObstacles_[i].apex_ + s * velocityObstacles_[i].side2_;
-
-                    if (candidate.position_.lengthSquared() < max_speed_ * max_speed_)
-                    {
-                        candidates_.insert(std::make_pair(
-                            (pref_velocity_ - candidate.position_).lengthSquared(),
-                            candidate));
-                    }
+                        velocityObstacles_[i].apex_ + s * velocityObstacles_[i].left_side;
+                    addToCandidateListIfValid(candidate);
                 }
             }
         }
     }
 
-    int optimal = -1;
+    double min_pref_speed = pref_velocity_.length() * min_pref_speed_multiplier;
 
-    for (std::pair<float, Candidate> candidate_pair : candidates_)
-    {
-        candidate  = candidate_pair.second;
-        bool valid = true;
+    // corresponds to the velocity obstacle that is furthest away when picking the new
+    // velocity that we might collide with
+    int optimal_furthest_away_obstacle = -1;
 
+    new_velocity_ = Vector();
+    // returns the first velocity obstacle in velocityObstacles_ that the given candidate
+    // point intersects. returns -1 if it doesn't intersect any velocity obstacle
+    auto firstIntersectingVelocityObstacle = [&](Candidate &c) -> int {
         for (int j = 0; j < static_cast<int>(velocityObstacles_.size()); ++j)
         {
-            if (j != candidate.velocityObstacle1_ && j != candidate.velocityObstacle2_ &&
-                (velocityObstacles_[j].side2_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) <
-                    0.0f &&
-                (velocityObstacles_[j].side1_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) >
-                    0.0f)
+            if ((j != c.velocityObstacle1_ && j != c.velocityObstacle2_ &&
+                 c.intersectsVelocityObstacle(velocityObstacles_[j])) ||
+                (c.position_.length() < min_pref_speed))
```

Comment:
I'm curious why we have the `c.position_.length() < min_pref_speed` case?

Reply:
this was to handle candidate velocities that didn't intersect any velocity obstacles but lower than the min speed threshold. this part is rewritten so that this function only has one purpose and the velocity checking is somewhere else